packages/edit-site/src/components/add-new-template/utils.js

```diff
@@ -169,9 +189,19 @@ const usePublicTaxonomies = () => {
 		[]
 	);
 	return useMemo( () => {
-		return taxonomies?.filter(
+		const publicTaxonomies = taxonomies?.filter(
 			( { visibility } ) => visibility?.publicly_queryable
 		);
+		taxonomyLabels = publicTaxonomies?.reduce(
+			( accumulator, { labels } ) => {
+				const singularName = labels.singular_name.toLowerCase();
+				accumulator[ singularName ] =
+					( accumulator[ singularName ] || 0 ) + 1;
+				return accumulator;
+			},
+			{}
+		);
+		return publicTaxonomies;
```

Comment:
We both know that in practice this "works", but it's a huge code smell.

Similarly to how async code can only be called by other async code¹, in Gutenberg we have a pure context and a "tainted" component-bound context, where React hooks must live. It's clear here that we have an increasing clash between those two contexts, with pieces that started out pure, like the entity configs, clashing with most of the dynamic data "trapped" in the impure hook-based context.

I think we need to make a decision on the architecture of the logic for adding templates. Some stream-of-consciousness ideas (take them with a grain of salt):

- Keep `needsUniqueIdentifier` pure but add a third argument named `namingConflicts` whose values are pre-computed near `useExtraTemplates`.
- Rearchitect away from this system of nested hooks and towards something more state-based (e.g. a large pure state selector) that can purely consume dynamic data from the core store.
- Rearchitect away from this system of pure per-entity configs and hardcode entity rules inside a more monolithic set of hooks.
- Leverage React context somehow to share the logic and pre-computations across `new-template` and `add-custom-template-modal`

* * *

_¹: And similarly to how, in other languages, monadic code (e.g. I/O monads, `unsafe` code blocks) can't be handled in "pure" contexts_